clientupdate/clientupdate.go

```diff
@@ -0,0 +1,888 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package clientupdate implements tailscale client update for all supported
+// platforms. This package can be used from both tailscaled and tailscale
+// binaries.
+package clientupdate
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"tailscale.com/net/tshttpproxy"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/must"
+	"tailscale.com/util/winutil"
+	"tailscale.com/version"
+	"tailscale.com/version/distro"
+)
+
+// Track is the Tailscale release track.
+type Track string
+
+const (
+	Current  = Track("")
+	Stable   = Track("stable")
+	Unstable = Track("unstable")
+)
+
+func versionToTrack(v string) (Track, error) {
+	_, rest, ok := strings.Cut(v, ".")
+	if !ok {
+		return "", fmt.Errorf("malformed version %q", v)
+	}
+	minorStr, _, ok := strings.Cut(rest, ".")
+	if !ok {
+		return "", fmt.Errorf("malformed version %q", v)
+	}
+	minor, err := strconv.Atoi(minorStr)
+	if err != nil {
+		return "", fmt.Errorf("malformed version %q", v)
+	}
+	if minor%2 == 0 {
+		return Stable, nil
+	}
+	return Unstable, nil
+}
+
+type updater struct {
+	UpdateArgs
+	update func() error
+}
+
+// UpdateArgs contains arguments needed to run an update.
+type UpdateArgs struct {
+	// Track is the release track to get updated version for. Defaults to
+	// Current, which means using the same track as the current running binary.
+	Track Track
+	// Version is the specific version to update to. This field may not be
+	// supported on all platforms (like Arch and Alpine Linuxes or FreeBSD).
+	// Mutually exclusive with Track.
```

Comment:
Call me an absolute maniac if you must — and indeed you must: Since Go doesn't have a way to express sum types, maybe this means we should just give in to the Dark Side and combine Track and Version into a single stringly typed field, with special constant values Current, Stable, and Unstable that we treat magically.

I can't believe I thought that thought and then typed it in. My Rust license is going to be revoked

Reply:
we still have separate flags on the CLI, but it does simplify the API of the library. I'll give it a shot and report back!

Reply:
updated, WDYT?